sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/expressions/expressions.scala

```diff
@@ -0,0 +1,208 @@
+    literal(value, internalLit.dataType)
+  }
+
+  def literal[T](value: T, dataType: DataType): LiteralValue[T] = LiteralValue(value, dataType)
+
+  def reference(name: String): NamedReference =
+    FieldReference(parser.parseMultipartIdentifier(name))
+
+  def apply(name: String, arguments: Array[Expression]): Transform = ApplyTransform(name, arguments)
+
+  def apply(name: String, arguments: Expression*): Transform = ApplyTransform(name, arguments)
+
+  def bucket(numBuckets: Int, columns: Array[String]): BucketTransform =
+    BucketTransform(literal(numBuckets, IntegerType), columns.map(reference))
+
+  def bucket(numBuckets: Int, columns: String*): BucketTransform =
+    BucketTransform(literal(numBuckets, IntegerType), columns.map(reference))
+
+  def identity(column: String): IdentityTransform = IdentityTransform(reference(column))
+
+  def year(column: String): YearTransform = YearTransform(reference(column))
+
+  def month(column: String): MonthTransform = MonthTransform(reference(column))
+
+  def date(column: String): DateTransform = DateTransform(reference(column))
+
+  def dateHour(column: String): DateHourTransform = DateHourTransform(reference(column))
+}
+
+/**
+ * Base class for simple transforms of a single column.
+ */
+private[sql] abstract class SingleColumnTransform(ref: NamedReference) extends Transform {
+
+  def reference: NamedReference = ref
+
+  override def references: Array[NamedReference] = Array(ref)
+
+  override def arguments: Array[Expression] = Array(ref)
+
+  override def describe: String = name + "(" + reference.describe + ")"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class BucketTransform(
+    numBuckets: Literal[Int],
+    columns: Seq[NamedReference]) extends Transform {
+
+  override val name: String = "bucket"
+
+  override def references: Array[NamedReference] = {
+    arguments
+        .filter(_.isInstanceOf[NamedReference])
+        .map(_.asInstanceOf[NamedReference])
+  }
+
+  override def arguments: Array[Expression] = numBuckets +: columns.toArray
+
+  override def describe: String = s"bucket(${arguments.map(_.describe).mkString(", ")})"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class ApplyTransform(
+    name: String,
+    args: Seq[Expression]) extends Transform {
+
+  override def arguments: Array[Expression] = args.toArray
+
+  override def references: Array[NamedReference] = {
+    arguments
+        .filter(_.isInstanceOf[NamedReference])
+        .map(_.asInstanceOf[NamedReference])
+  }
+
+  override def describe: String = s"$name(${arguments.map(_.describe).mkString(", ")})"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class IdentityTransform(
+    ref: NamedReference) extends SingleColumnTransform(ref) {
+  override val name: String = "identity"
+  override def describe: String = ref.describe
+}
+
+private[sql] final case class YearTransform(
```

Comment:
why it's private? Do we expect the data source implementation to catch it?

Reply:
The Spark implementations private because catalogs should not rely on specific implementation classes. The API is an interface and catalogs must rely on that.

Reply:
say I'm a table catalog developer, if I want to support `PARTITION BY year(ts)`, I need to call `Transform.name` and `Transform. arguments` to check if it's a year transform?

Reply:
`Transform.name` will be `year`. `Transform.arguments` will contain the column, as will `Transform.references`.